doc/arch/adr-004-transition-mainstream-to-publishing-api-index.md

```diff
@@ -0,0 +1,107 @@
+# Decision record: Transition mainstream formats to a Publishing API derived search index
+
+Date: 2017-08-24
+
+## Definitions
+
+Throughout this document, `format` refers to the Rummager field `format`, not the document type provided by Publishing API (`content_store_document_type`).
+
+Every `content_store_document_type` is mapped to a single `format`.
+
+`mainstream` is used as an example of a search index we want to retire. This document should also apply to the `government` and `detailed` indexes.
+
+## Context
+### We're building a new search index using publishing API data
+We're currently replacing the existing search indexes (`mainstream`, `government`, `detailed`) with a single index (`govuk`), which is derived from Publishing API.
+
+The existing indexes are populated from [at least 10 different applications](https://docs.publishing.service.gov.uk/apps.html).
+
+This change should improve the quality of the data our search system relies on:
+
+- all formats are handled in a consistent manner
+- we can update documents after any publishing event (publishing, withdrawing, unpublishing, tagging)
+- it should be easier to rebuild if we need to revert to an old backup
+
+### Lifecycle of our search indexes
+Rumamger's search indexes are long lived, but content is regularly
+updated to derive the `popularity` field from recent pageviews.
+As of [ADR 003: Popularity updating without index locks](adr-003-popularity-updating-without-index-locks.md),
+this uses updates, rather than rebuilding the search index every time.
+
+There is a separate task in place to reindex the content currently in the search indexes with [zero downtime](https://www.elastic.co/guide/en/elasticsearch/guide/current/index-aliases.html).
+This is something we need to do after adding new fields or otherwise changing
+the Elasticsearch mappings, for those fields to work properly.
+
+Note that neither of these mechanisms add or remove documents from the search index.
+
+- If an edition is ever published, and the search index doesn't get updated, it stays out of date until the next time the document is updated
+
+- If an edition is unpublished without Rummager being notified, it stays in the search index forever
+
+Long term, we'd like to be able to easily rebuild the whole `govuk` index
+from publishing API, to avoid these kind of problems, but we aren't planning to do this right now.
+
+### Immediate needs
+
+We need to be able to populate the `govuk` index with all the documents that
+already exist, so we can start using this index in the Search API.
+
+We also want to be able to reindex everything that has been published within a
+short period of time (1 or 2 days), so that we can easily recover the index
+from backups.
```

Comment:
without using backups? we will be rebuilding from scratch after all. Also, ES will take approx 1-2hrs to reinsert the data so maybe we should be more optimistic with the time estimate

Reply:
I'm assuming that we don't need to do the "from scratch" part. This simplifies the implementation a bit as we can be less concerned about performance, and don't have to coordinate all the index management stuff (working out when to actually switch aliases).

I think we don't actually need to do a full rebuild from scratch now though, and the big win is just replaying the last N days.

Reply:
I've pushed a new commit with a separate section for "problems we're not solving", so hopefully it's a bit clearer now.